examples/timestamp-queries/src/main.rs

```diff
@@ -0,0 +1,484 @@
+            );
+        }
+    }
+}
+
+impl Queries {
+    fn new(device: &wgpu::Device, num_queries: u64) -> Self {
+        Queries {
+            set: device.create_query_set(&wgpu::QuerySetDescriptor {
+                label: Some("Timestamp query set"),
+                count: num_queries as _,
+                ty: wgpu::QueryType::Timestamp,
+            }),
+            resolve_buffer: device.create_buffer(&wgpu::BufferDescriptor {
+                label: Some("query resolve buffer"),
+                size: std::mem::size_of::<u64>() as u64 * num_queries,
+                usage: wgpu::BufferUsages::COPY_SRC | wgpu::BufferUsages::QUERY_RESOLVE,
+                mapped_at_creation: false,
+            }),
+            destination_buffer: device.create_buffer(&wgpu::BufferDescriptor {
+                label: Some("query dest buffer"),
+                size: std::mem::size_of::<u64>() as u64 * num_queries,
+                usage: wgpu::BufferUsages::COPY_DST | wgpu::BufferUsages::MAP_READ,
+                mapped_at_creation: false,
+            }),
+            num_queries,
+            next_unused_query: 0,
+        }
+    }
+
+    fn resolve(&self, encoder: &mut wgpu::CommandEncoder) {
+        encoder.resolve_query_set(
+            &self.set,
+            // TODO(https://github.com/gfx-rs/wgpu/issues/3993): Musn't be larger than the number valid queries in the set.
+            0..self.next_unused_query as u32,
+            &self.resolve_buffer,
+            0,
+        );
+        encoder.copy_buffer_to_buffer(
+            &self.resolve_buffer,
+            0,
+            &self.destination_buffer,
+            0,
+            self.resolve_buffer.size(),
+        );
+    }
+
+    fn wait_for_results(&self, device: &wgpu::Device) -> Vec<u64> {
+        self.destination_buffer
+            .slice(..)
+            .map_async(wgpu::MapMode::Read, |_| ());
+        device.poll(wgpu::Maintain::Wait);
+
+        let timestamps = {
+            let timestamp_view = self
+                .destination_buffer
+                .slice(..(std::mem::size_of::<u64>() as wgpu::BufferAddress * self.num_queries))
+                .get_mapped_range();
+            bytemuck::cast_slice(&timestamp_view).to_vec()
+        };
+
+        self.destination_buffer.unmap();
+
+        timestamps
+    }
+}
+
+async fn run() {
+    // Instantiates instance of WebGPU
+    let backends = wgpu::util::backend_bits_from_env().unwrap_or_else(wgpu::Backends::all);
+    let instance = wgpu::Instance::new(wgpu::InstanceDescriptor {
+        backends,
+        dx12_shader_compiler: wgpu::Dx12Compiler::default(),
+    });
+
+    // `request_adapter` instantiates the general connection to the GPU
+    let adapter = instance
+        .request_adapter(&wgpu::RequestAdapterOptions::default())
+        .await
+        .expect("Failed to request adapter.");
+
+    // Check timestamp features.
+    if adapter.features().contains(wgpu::Features::TIMESTAMP_QUERY) {
```

Comment:
Let's cache the features call in a variable, it's not expensive, but not just a getter.